man/mscan.1
.Dd January 2, 2020
.Dt MSCAN 1
.Os
.Sh NAME
.Nm mscan
.Nd generate one-line message summaries
.Sh SYNOPSIS
.Nm
.Op Fl I
.Op Fl n
.Op Fl v
.Op Fl f Ar format
.Ar msgs\ ...
.Sh DESCRIPTION
.Nm
prints a one line summary for each message.
.Po
See
.Xr mmsg 7
for the message argument syntax.
.Pc
.Pp
If no
.Ar msgs
are specified,
.Nm
reads filenames from the standard input,
or scans the mails in the current sequence if used interactively.
.Pp
By default,
.Nm
will spawn a pager on its output when used interactively.
See
.Sx ENVIRONMENT
for details on how to control this behaviour.
.Pp
The default
.Ar format
is
.Pp
.Dl %c%u%r %-3n %10d %17f %t %2i%s
.Pp
that is, for each message,
.Nm
prints relevant flags,
the sequence number
.Pq if applicable ,
the date,
the originator,
and the subject of the message
.Pq possibly indented .
A default
.Nm
format may be specified in the user's
.Xr mblaze-profile 5 .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl I
Force ISO date output,
even for
.Sq Cm "%d" .
.It Fl n
Only print message numbers
.Pq or filenames, if the message is not in the current sequence .
.It Fl v
Print how many messages were scanned to standard error.
.It Fl f Ar format
Format according to the string
.Ar format ,
inspired by
.Xr printf 3 .
.Pp
The following formatting codes may be used
.Po
.Ar wd
overrides the default width
.Pc :
.Bl -tag -width Ds
.It Cm \en
Newline.
.It Cm \et
Tab.
.It Cm \&%%
A plain
.Sq Li \&% .
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "b"
Human-readable size of the message
.Pq in kilobytes .
.It Cm "%c"
A
.Sq Li >
on the current message,
otherwise a blank.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "d"
Adaptive date of the message.
When
.Ar wd
is greater or equal to 19,
the timestamp is shown in dottime format
.Pq Lk https://dotti.me/ .
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "D"
ISO date of the message
.Pq year, month, day .
When
.Ar wd
is greater or equal to 16,
the hour-minute timestamp will also be shown.
When
.Ar wd
is greater or equal to 19,
seconds will also be shown.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "f"
The
.Sq Li From\&:
.Po
or
.Sq Li To\&: ,
if the message is from us
.Pc .
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "F"
The maildir folder the message resides in.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "i"
.Ar wd
.Pq default: 1
spaces per indentation depth in the thread tree.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "I"
The
.Sq Li Message\&-ID\&:
of the message.
.It Cm "%M"
The raw maildir flags of the message.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "n"
The number of the message in the current sequence.
.It Cm "%r"
A
.Sq Li \&-
on a replied-to message, or a
.Sq Li \&:
on a forwarded message, or a blank.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "R"
The filename of the message.
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "s"
The subject of the message
.Pq defaults to remaining width .
.It Cm "%" Ns Oo Ar wd Oc Ns Cm "S"
The subject of the message
.Pq defaults to remaining width ,
with leading
.Sq Li Re\&: ,
.Sq Li Fwd\&:
etc. stripped.
.It Cm "%t"
A
.Sq Li \&>
if you are in
.Sq Li To\&: ,
a
.Sq Li \&+
if you are in
.Sq Li Cc\&: ,
a
.Sq Li \&:
if you are in
.Sq Li Resent\&-To\&: ,
or a blank.
.It Cm "%u"
An
.Sq Li \&*
on a flagged message, a
.Sq Li \&.
on an unseen message, an
.Sq Li x
on a trashed message, or a blank.
.El
.El
.Sh MESSAGE FLAGS
.Bl -tag -width 2n -compact
.It Li \&>
The current message
.It Li \&.
An unseen message
.It Li x
A trashed message
.It Li \&*
A flagged message
.It Li \&-
A replied-to message
.El
.Sh SUBJECT FLAGS
.Bl -tag -width 2n -compact
.It Li \&>
You are in
.Sq Li To\&:
.It Li \&+
You are in
.Sq Li Cc\&:
.It Li \&:
You are in
.Sq Li Resent\&-To\&:
.El
.Sh ENVIRONMENT
.Bl -tag -width MBLAZE_PAGER
.It Ev MBLAZE_PAGER
Any non-empty value of the environment variable
.Ev MBLAZE_PAGER
is used instead of the standard pagination program, specified in
.Ev PAGER .
When empty, or set to
.Sq Ic cat ,
no pager is spawned.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr mblaze-profile 5 ,
.Xr mmsg 7
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## 0.5.1 (2019-03-03)

* mdeliver: preserve mtime in mrefile
* mdirs: add -0 to separate folders by NUL characters
* Fixes for buffer-overflows, found by fuzzing.
* Fixes for memleaks.

## 0.5 (2019-02-09)

* New tool msearch to wrap several mail indexers.
* New zsh completion _mblaze.
* mnext/mprev were removed (you can call `mless +`/`mless -`).
* The GnuPG tools in contrib/ now use gpg2.
* mshow exits with error if it could not extract all attachments
* mrep: add -noquote to disable quoting the message replied to
* mdeliver: keep permissions of messages
* mcom: aborting the editor is now more like delete than cancel
* mcom: add -send to send directly without editing
* mcom: check if mail is formatted sensibly
* mpick: new flag -v for statistics
* mscan: new flag -v for statistics
* magrep: add -h, which is like -p but doesn't print the file name
* mscan: prioritize displaying trashed mail over other markers
* mpick: fix off-by-one in expression parsing
* Many bug fixes

## 0.4 (2018-08-15)

* New tool mrefile to move or copy messages.
* contrib/mp7m: add application/pkcs7-mime decoder
* mcom: allow setting arbitrary headers from command line
* mcom: add -body to prepopulate drafts
* mcom: mark drafts as seen after sending
* mcom: update flags after mrep/mbnc/mfwd
* mshow: add -B to decode MIME parameters in broken mails
* magrep: add -l flag
* Many bug fixes.

## 0.3.2 (2018-02-13)

* magrep: add *:REGEX to search in any header
* Fix of a buffer overflow in blaze822_multipart.
* Small bug fixes.
* Many documentation improvements by Larry Hynes.

## 0.3.1 (2018-01-30)

* mless: support $NO_COLOR
* mcolor: support $NO_COLOR
* blaze822.h: ensure PATH_MAX is defined
* Improved documentation.
* Many fixes for address parser.

## 0.3 (2018-01-12)

* New tool mflow to reformat format=flowed plain text mails.
* New tool mbnc to bounce mails (send original mail to someone else).
* mshow filters can output raw text now, e.g. for HTML rendering with colors.
* mrep can quote mail that doesn't have a plain text part.
* mcom runs mmime when deemed necessary.
* mhdr can extract MIME parameters.
* New contrib: mrecode
* New contrib: mraw
* mshow regards non-MIME mails as MIME mails with one part now.
* mshow -F to disable MIME filters.
* mpick supports negations now.
* msed can remove headers depending on their value.
* Improved UTF-8 parsing.
* Improved documentation.
* Numerous bug fixes and portability fixes.

## 0.2 (2017-07-17)

* New sequence syntax `m:+n` for `n` messages after message `m`.
* Threading shortcuts `=`, `_`, `^` for `.=`, `._`, `.^`.
* Sequence related errors are now reported.
* minc and mlist normalize slashes in paths.
* mfwd now generates conforming message/rfc822 parts.
* mthread can add optional folders (e.g. your outbox) to resolve message ids.
* mcom now adds Date: just before sending or cancelling the mail.
* VIOLATIONS.md documents how mblaze works with certain common mistakes.
* Full documentation revamp by Larry Hynes.
* Fix rare crash looking for mail body.
* Numerous small bug and portability fixes.

## 0.1 (2017-06-24)

* Initial release


mscan.c
#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 700
#endif

#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <unistd.h>
#include <wchar.h>

#include "blaze822.h"
#include "u8decode.h"

static int cols;
static wchar_t replacement = L'?';
static char *cur;

static char *aliases[32];
static int alias_idx;

static int Iflag;
static int nflag;
static int vflag;
static int curyear;
static time_t now;
static char default_fflag[] = "%c%u%r %-3n %10d %17f %t %2i%s";
static char *fflag = default_fflag;

ssize_t
u8putstr(FILE *out, char *s, ssize_t l, int pad)
{
	ssize_t ol = l;

	while (*s && l > 0) {
		if (*s == '\t')
			*s = ' ';

		if ((unsigned)*s < 32 || *s == 127) {  // C0
			fprintf(out, "%lc", (wint_t)(*s == 127 ? 0x2421 : 0x2400+*s));
			s++;
			l--;
		} else {
			uint32_t c;
			int r = u8decode(s, &c);
			if (r < 0) {
				r = 1;
				l--;
				fprintf(out, "%lc", (wint_t)replacement);
			} else {
				int w = wcwidth((wchar_t)c);
				if (w < 0)
					w = 2;   /* assume worst width */
				l -= w;
				if (l >= 0)
					fwrite(s, 1, r, out);
			}
			s += r;
		}
	}
	if (pad)
		while (l-- > 0)
			putc(' ', out);

	if (l < 0)
		l = 0;
	return ol - l;
}

int
itsme(char *v)
{
	int i;

	char *disp, *addr;
	while ((v = blaze822_addr(v, &disp, &addr)))
		for (i = 0; addr && i < alias_idx; i++)
			if (strcmp(aliases[i], addr) == 0)
				return 1;
	return 0;
}

static int init;

void
numline(char *file)
{
	if (!init) {
		// delay loading of the seq until we need to scan the first
		// file, in case someone in the pipe updated the map before
		char *seq = blaze822_seq_open(0);
		blaze822_seq_load(seq);
		cur = blaze822_seq_cur();
		init = 1;
	}

	while (*file == ' ' || *file == '\t')
		file++;

	long lineno = blaze822_seq_find(file);
	if (lineno)
		printf("%ld\n", lineno);
	else
		printf("%s\n", file);
}

static char *
fmt_date(struct message *msg, int w, int iso)
{
	static char date[32];
	char *v;

	if (!msg)
		return "";

	v = blaze822_hdr(msg, "date");
	if (!v)
		return "(unknown)";
	time_t t = blaze822_date(v);
	if (t == -1)
		return "(invalid)";

	struct tm *tm;
	tm = localtime(&t);

	if (iso) {
		if (w >= 19)
			strftime(date, sizeof date, "%Y-%m-%d %H:%M:%S", tm);
		else if (w >= 16)
			strftime(date, sizeof date, "%Y-%m-%d %H:%M", tm);
		else
			strftime(date, sizeof date, "%Y-%m-%d", tm);
	} else if (w >= 19) {
		// https://dotti.me/

		tm = gmtime(&t);
		strftime(date, sizeof date, "%Y-%m-%dT%H\xc2\xb7%M", tm);
		char *d = date + strlen(date);

		char *e = v + strlen(v);
		while (e > v && (*e != '+' && *e != '-'))
			e--;

		if (isdigit(e[1]) && isdigit(e[2]) && isdigit(e[3]) && isdigit(e[4])) {
			*d++ = e[0];
			*d++ = e[1];
			*d++ = e[2];
			if (e[3] != '0' || e[4] != '0') {
				*d++ = ':';
				*d++ = e[3];
				*d++ = e[4];
			}
			*d = 0;
		} else {
			*d++ = '+';
			*d++ = '0';
			*d++ = '0';
			*d = 0;
		}
	} else if (w < 10) {
		if (tm->tm_year != curyear)
			strftime(date, sizeof date, "%b%y", tm);
		else if (t > now || now - t > 86400)
			strftime(date, sizeof date, "%d%b", tm);
		else
			strftime(date, sizeof date, "%H:%M", tm);
	} else {
		if (tm->tm_year != curyear)
			strftime(date, sizeof date, "%Y-%m-%d", tm);
		else if (t > now || now - t > 86400)
			strftime(date, sizeof date, "%a %b %e", tm);
		else
			strftime(date, sizeof date, "%a  %H:%M", tm);
	}

	return date;
}

static char *
fmt_subject(struct message *msg, char *file, int strip)
{
	static char subjdec[100];
	char *subj = "(no subject)";
	char *v;

	if (!msg) {
		snprintf(subjdec, sizeof subjdec, "\\_ %s", file);
		return subjdec;
	}

	if ((v = blaze822_hdr(msg, "subject")))
		subj = v;

	blaze822_decode_rfc2047(subjdec, subj, sizeof subjdec - 1, "UTF-8");

	if (strip) {
		size_t i;
		for (i = 0; subjdec[i]; ) {
			if (subjdec[i] == ' ') {
				i++;
				continue;
			} else if (strncasecmp("re:", subjdec+i, 3) == 0 ||
				   strncasecmp("aw:", subjdec+i, 3) == 0) {
				i += 3;
				continue;
			} else if (strncasecmp("fwd:", subjdec+i, 4) == 0) {
				i += 4;
				continue;
			}
			break;
		}
		return subjdec + i;
	}

	return subjdec;
}

static char *
fmt_from(struct message *msg)
{
	static char fromdec[256];
	char *from = "(unknown)";
	char *v, *w;

	if (!msg)
		return "";

	if ((v = blaze822_hdr(msg, "from"))) {
		blaze822_decode_rfc2047(fromdec, v, sizeof fromdec - 1, "UTF-8");
		fromdec[sizeof fromdec - 1] = 0;
		from = fromdec;

		if (itsme(fromdec) && ((w = blaze822_hdr(msg, "to")))) {
			memcpy(fromdec, "TO:", 4);
			blaze822_decode_rfc2047(fromdec + 3, w,
			    sizeof fromdec - 1 - 3, "UTF-8");
			from = fromdec;
		} else {
			char *disp, *addr;
			blaze822_addr(fromdec, &disp, &addr);
			if (disp)
				from = disp;
			else if (addr)
				from = addr;
		}
	}

	return from;
}

static char *
fmt_to_flag(struct message *msg)
{
	char *v;

	if (!msg || !alias_idx)
		return " ";

	if ((v = blaze822_hdr(msg, "to")) && itsme(v))
		return ">";
	else if ((v = blaze822_hdr(msg, "cc")) && itsme(v))
		return "+";
	else if ((v = blaze822_hdr(msg, "resent-to")) && itsme(v))
		return ":";
	else if ((v = blaze822_hdr(msg, "from")) && itsme(v))
		return "<";
	else
		return " ";
}

static ssize_t
print_human(intmax_t i, int w)
{
	double d = i / 1024.0;
	const char *u = "\0\0M\0G\0T\0P\0E\0Z\0Y\0";
	while (d >= 1024) {
		u += 2;
		d /= 1024.0;
	}

	if (d < 1.0)
		return printf("%*.2f", w, d);
	else if (!*u)
		return printf("%*.0f", w, d);
	else if (d < 10.0)
		return printf("%*.1f%s", w-1, d, u);
	else
		return printf("%*.0f%s", w-1, d, u);
}

void
oneline(char *file)
{
	if (!init) {
		// delay loading of the seq until we need to scan the first
		// file, in case someone in the pipe updated the map before
		char *seq = blaze822_seq_open(0);
		blaze822_seq_load(seq);
		cur = blaze822_seq_cur();
		init = 1;
	}

	int indent = 0;
	while (*file == ' ' || *file == '\t') {
		indent++;
		file++;
	}

	struct message *msg = blaze822(file);
	char *flags = msg ? strstr(file, ":2,") : 0;
	if (!flags)
		flags = "";
	else
		flags += 3;

	int wleft = cols;

	char *f;
	for (f = fflag; *f; f++) {
		if (*f == '\\') {
			f++;
			switch (*f) {
			case 'n': putchar('\n'); wleft = cols; break;
			case 't': putchar('\t'); wleft -= (8 - wleft % 8); break;
			default:
				putchar('\\'); wleft--;
				putchar(*f); wleft--;
			}
			continue;
		}
		if (*f != '%') {
			putchar(*f);
			wleft--;
			continue;
		}
		f++;

		int w = 0;
		if ((*f >= '0' && *f <= '9') || *f == '-') {
			errno = 0;
			char *e;
			w = strtol(f, &e, 10);
			if (errno != 0)
				w = 0;
			else
				f = e;
		}

		if (!*f)
			break;

		switch (*f) {
		case '%':
			putchar('%');
			wleft--;
			break;
		case 'c':
			if (cur && strcmp(cur, file) == 0)
				putchar('>');
			else
				putchar(' ');
			wleft--;
			break;
		case 'u':  // unseen
			if (strchr(flags, 'T'))
				putchar('x');
			else if (strchr(flags, 'F'))
				putchar('*');
			else if (msg && !strchr(flags, 'S'))
				putchar('.');
			else
				putchar(' ');
			wleft--;
			break;
		case 'r':  // replied
			if (strchr(flags, 'R'))
				putchar('-');
			else if (strchr(flags, 'P'))
				putchar(':');
			else
				putchar(' ');
			wleft--;
			break;
		case 't':  // to-flag
			wleft -= printf("%s", fmt_to_flag(msg));
			break;
		case 'M':  // raw Maildir flags
			if (!w) w = -3;
			wleft -= printf("%*s", w, flags);
			break;
		case 'n':
			{
				long lineno = msg ? blaze822_seq_find(file) : 0;

				if (lineno)
					wleft -= printf("%*ld", w, lineno);
				else
					wleft -= printf("%*s", w, "");
			}
			break;
		case 'd':
		case 'D':
			if (!w) w = 10;
			wleft -= printf("%*s", w,
			    fmt_date(msg, w, Iflag || *f == 'D'));
			break;
		case 'f':
			if (w < 0)
				w += wleft;
			if (w)
				wleft -= u8putstr(stdout,
				    fmt_from(msg), w, 1);
			else
				wleft -= u8putstr(stdout,
				    fmt_from(msg), wleft, 0);
			break;
		case 'i':
			{
				int z;
				if (!w)
					w = 1;

				if (indent >= 10) {
					wleft -= printf("..%2d..", indent);
					indent = 10 - 6/w;
				}

				for (z = 0; z < w*indent; z++)
					putchar(' ');
				wleft -= w*indent;
			}
			break;
		case 's':
		case 'S':
			if (w < 0)
				w += wleft;
			if (w)
				wleft -= u8putstr(stdout,
				    fmt_subject(msg, file, *f == 'S'), w, 1);
			else
				wleft -= u8putstr(stdout,
				    fmt_subject(msg, file, *f == 'S'), wleft, 0);
			break;
		case 'b':
			{
				struct stat st;
				if (msg) {
					if (stat(file, &st) != 0)
						st.st_size = 0;
					wleft -= print_human(st.st_size, w);
				} else {
					wleft -= printf("%.*s", w, "");
				}
			}
			break;
		case 'F':
			{
				char *e = file + strlen(file);
				while (file < e && *e != '/')
					e--;
				e--;
				while (file < e && *e != '/')
					e--;
				while (file < e && *e == '/')
					e--;
				char *b = e;
				e++;
				while (file < b && *b != '/')
					b--;
				if (*b == '/')
					b++;
				if (*b == '.')
					b++;
				if (w) {
					if (w < 0)
						w = -w;
					wleft -= printf("%*.*s",
					    -w, (int)(e-b < w ? e-b : w), b);
				} else {
					wleft -= printf("%.*s", (int)(e-b), b);
				}
			}
			break;
		case 'R':
			if (w)
				wleft -= printf("%*.*s", w, w, file);
			else
				wleft -= printf("%s", file);
			break;
		case 'I':
			{
				char *m = msg ? blaze822_hdr(msg, "message-id") : 0;
				if (!m)
					m = "(unknown)";
				if (w)
					wleft -= printf("%*.*s", w, w, m);
				else
					wleft -= printf("%s", m);
			}
			break;
		default:
			putchar('%');
			putchar(*f);
			wleft -= 2;
		}
	}

	printf("\n");

	blaze822_free(msg);
}

int
main(int argc, char *argv[])
{
	pid_t pager_pid = -1;

	int c;
	while ((c = getopt(argc, argv, "If:nv")) != -1)
		switch (c) {
		case 'I': Iflag++; break;
		case 'f': fflag = optarg; break;
		case 'n': nflag = 1; break;
		case 'v': vflag = 1; break;
		default:
			fprintf(stderr, "Usage: mscan [-Inv] [-f format] [msgs...]\n");
			exit(1);
		}

	if (nflag) {
		if (argc == optind && isatty(0))
			blaze822_loop1(":", numline);
		else
			blaze822_loop(argc-optind, argv+optind, numline);
		return 0;
	}

	now = time(0);
	struct tm *tm = localtime(&now);
	curyear = tm->tm_year;

	setlocale(LC_ALL, "");  // for wcwidth later
	if (wcwidth(0xfffd) > 0)
		replacement = 0xfffd;

	struct winsize w;
	int ttyfd = open("/dev/tty", O_RDONLY | O_NOCTTY);
	if (ttyfd >= 0 && ioctl(ttyfd, TIOCGWINSZ, &w) == 0) {
		cols = w.ws_col;

		char *pg;
		pg = getenv("MBLAZE_PAGER");
		if (!pg)
			pg = getenv("PAGER");
		if (pg && *pg && strcmp(pg, "cat") != 0) {
			pager_pid = pipeto(pg);
			if (pager_pid < 0)
				fprintf(stderr,
				    "mscan: spawning pager '%s': %s\n",
				    pg, strerror(errno));
		}
	}
	if (ttyfd >= 0)
		close(ttyfd);
	if (getenv("COLUMNS"))
		cols = atoi(getenv("COLUMNS"));
	if (cols <= 40)
		cols = 80;

	char *f = blaze822_home_file("profile");
	struct message *config = blaze822(f);

	if (config) {
		char *v, *d, *a;
		if ((v = blaze822_hdr(config, "local-mailbox")))
			while (alias_idx < (int)(sizeof aliases / sizeof aliases[0]) &&
			    (v = blaze822_addr(v, &d, &a)))
				if (a)
					aliases[alias_idx++] = strdup(a);
		if ((v = blaze822_hdr(config, "alternate-mailboxes")))
			while (alias_idx < (int)(sizeof aliases / sizeof aliases[0]) &&
			    (v = blaze822_addr(v, &d, &a)))
				if (a)
					aliases[alias_idx++] = strdup(a);
		if ((v = blaze822_hdr(config, "scan-format")))
			if (fflag == default_fflag)  // NB. ==
				fflag = v;
	}

	long i;
	if (argc == optind && isatty(0))
		i = blaze822_loop1(":", oneline);
	else
		i = blaze822_loop(argc-optind, argv+optind, oneline);

	if (pager_pid > 0)
		pipeclose(pager_pid);

	if (vflag)
		fprintf(stderr, "%ld mails scanned\n", i);

	return 0;
}
